icecandidate_event: null, empty string, variables in examples (#27392)
mdn/content: files/en-us/web/api/rtcpeerconnection/icecandidate_event/index.md

{{APIRef("WebRTC")}}

An **`icecandidate`** event is sent to an {{domxref("RTCPeerConnection")}} when:

- An {{domxref("RTCIceCandidate")}} has been identified and added to the local peer by a call to {{domxref("RTCPeerConnection.setLocalDescription()")}},

> - Every {{domxref("RTCIceCandidate")}} correlated with a particular Username Fragment and Password combination (a **generation**) has been so identified and added, and

Reviewer: ```suggestion
- Every {{domxref("RTCIceCandidate")}} correlated with a particular username fragment and password combination (a **generation**) has been so identified and added, and
```

{{APIRef("WebRTC")}}

> An **`icecandidate`** event is sent to an {{domxref("RTCPeerConnection")}} when A) an {{domxref("RTCIceCandidate")}} has been identified and added to the local peer by a call to {{domxref("RTCPeerConnection.setLocalDescription()")}}, B) every {{domxref("RTCIceCandidate")}} correlated with a particular Username Fragment and Password combination (a **generation**) has been so identified and added, and C) all ICE gathering on all transports is complete. In the first two cases, the event handler should transmit the candidate to the remote peer over the signaling channel so the remote peer can add it to its set of remote candidates.

Reviewer: - I would bullet the list. Generally if a list is not readable just with commas, then bullets make it more readable that trying to stay inline and using the "A, B, C" markers. That's an opinion, not a rule.
- Are Username Fragment and Password proper nouns in this case or should they be lower case.

Author: The Username Fragment and Password combination is specified with capital letters in RFC 8838 Section 2 in the definition of "generation".

Reviewer: We're not writing the spec. They should be lower cased.

_A {{domxref("RTCPeerConnectionIceEvent")}} being an {{domxref("Event")}}, this event also implements these properties_.

- {{domxref("RTCPeerConnectionIceEvent.candidate")}} {{ReadOnlyInline}}

>   - : Contains the {{domxref("RTCIceCandidate")}} containing the candidate associated with the event, or the empty string if this event indicates that there are no further candidates to come.

Reviewer: Contains the ... containing is a bit ugly.

 ```suggestion
  - : The {{domxref("RTCIceCandidate")}} containing the candidate associated with the event, or the empty string if this event indicates that there are no further candidates to come.
```

Author: Contains is a bit ugly.  Returns suggests it's a function and it's not.  As a property, it actually indicates, so good point, and I will change it to indicates.

## Event properties

_A {{domxref("RTCPeerConnectionIceEvent")}} being an {{domxref("Event")}}, this event also implements the following property_.

- {{domxref("RTCPeerConnectionIceEvent.candidate")}} {{ReadOnlyInline}}

>   - : Indicates the {{domxref("RTCIceCandidate")}} containing the candidate associated with the event. This will be the empty string if the event indicates that there are no further candidates to come in this **generation**, or `null` if all ICE gathering on all transports is complete.

Reviewer: ```suggestion
  - : Indicates the {{domxref("RTCIceCandidate")}} containing the candidate associated with the event.
    This will be the empty string if the event indicates that there are no further candidates to come in this **generation**, or `null` if all ICE gathering on all transports is complete.
```

## Event properties

> _A {{domxref("RTCPeerConnectionIceEvent")}} being an {{domxref("Event")}}, this event also implements the following property_.

Reviewer: FYI only, I would be generally fine to leave as is, even if there is only one property. No need to revert of course.

- {{domxref("RTCPeerConnectionIceEvent.candidate")}} {{ReadOnlyInline}}

>   - : Indicates the {{domxref("RTCIceCandidate")}} containing the candidate associated with the event.
>   - : The empty string if this event indicates that there are no further candidates to come in this **generation**.
>   - : `null` if all ICE gathering on all transports is complete.

Reviewer: The syntax for multiple definitions is not supported, so does not render: https://pr27392.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/RTCPeerConnection/icecandidate_event#event_properties

If you're OK with the version below, I think we can merge (or counter propose):

```suggestion
  - : Indicates the {{domxref("RTCIceCandidate")}} containing the candidate associated with the event.
     This will be the empty string if the event indicates that there are no further candidates to come in this **generation**, or `null` if all ICE gathering on all transports is complete.
```

_A {{domxref("RTCPeerConnectionIceEvent")}} being an {{domxref("Event")}}, this event also implements these properties_.

- {{domxref("RTCPeerConnectionIceEvent.candidate")}} {{ReadOnlyInline}}

>   - : Indicates the {{domxref("RTCIceCandidate")}} containing the candidate associated with the event, or the empty string if this event indicates that there are no further candidates to come.

Reviewer: So this is a general section, so should cover all the cases. Feel free to counter propose some words. You understand this better than me. 

```suggestion
  - : Indicates the {{domxref("RTCIceCandidate")}} containing the candidate associated with the event,
     This is sent with the empty string if there are no further candidates to come, and `null` if all ICE transports have finished gathering candidates (ICE gathering is complete).
```

### Indicating that ICE gathering is complete

> Once all ICE transports have finished gathering candidates and the value of the {{domxref("RTCPeerConnection")}} object's {{domxref("RTCPeerConnection.iceGatheringState", "iceGatheringState")}} has made the transition to `complete`, an `icecandidate` event is sent with the value of `candidate` set to `null`.

Reviewer: Above you changed the return value of candidate to "the empty string" from `null`. Which is it? Null is used in a number of place below too, so all need to be consistent.

Reviewer: And if it can be null in some cases and empty string in another, we should state both cases.

Author: Once all ICE transports have finished, in order to indicate the transition to complete, it must be null and cannot be the empty string.  I did change it earlier to "the empty string" because the documentation was talking about the end of a generation and not the transition to complete.  If it were the empty string, it would not be complete, as this implies that there is another Username Fragment and Password combination that may provide more candidates.

So you're right, it can be either, and the difference is described in the section you included here ("Once all ICE...") and the previous section where it says 'whose candidate string is empty ("").'  So we are stating both cases.

Reviewer: Thanks. This section is correct, but we should mention all cases in the properties section, including `null`. I'll propose some words.

pc.addEventListener(
  "icecandidate",
  (ev) => {

>     if (ev.candidate !== null) {

Reviewer: Null or empty string?

Author: It should pass the empty string because the candidate was passed to indicate that all the candidates correlated with a particular Username Fragment and Password combination have been passed, so the next one that that gets passed is from a different combination.
